### DISCUSSION TITLE
Will the activation be reordered when act_order is true for gptq?

### DISCUSSION THREAD

**Discussion by frankxyy** [turboderp/exllamav2#208] [General]
If the weight is reordered to let the group_idx be sequential among the columns. The activation order seems to be affected. So does exllama reordered the activation order during inference?

**Comment by turboderp** [comment]
Yes, it sorts columns of the linear layer by their group index, which allows act-order matrices to be processed the same way as non act-order matrices. Then at inference time, the input to the linear layer is reordered accordingly (same permutation) and you end up with the same output: WX = P_col(W) P_row(X), or XW = P_col(x) P_row(W) depending on how you look at it. 

This way you only have to apply a fixed permutation to the input vector, which turns out to be much more efficient than the random access to load a different set of group parameters for every individual column.
